File: hive-storage-handler/src/main/scala/net/bitnine/agens/hive/livy/PiJob.scala
package net.bitnine.agens.hive.livy

import org.apache.livy.{Job, JobContext}

class PiJob(val slices: Int) extends Job[java.lang.Double] {

	val samples = Math.min(100000L * slices, Integer.MAX_VALUE).toInt

	override def call(jc: JobContext): java.lang.Double = {
		// Pi Estimation
		// https://spark.apache.org/examples.html

		val count = jc.sc.parallelize(1 to samples).filter { _ =>
			val x = math.random
			val y = math.random
			x*x + y*y < 1
		}.count()

		val pi = 4.0 * count / samples
		pi
	}

}

File: livy-test/src/main/java/net/bitnine/agens/livytest/java/PiJob.java
package net.bitnine.agens.livytest.java;

import java.util.ArrayList;
import java.util.List;

import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.Function2;

import org.apache.livy.*;

class PiJob implements
        Job<Double>,
        Function<Integer, Integer>,
        Function2<Integer, Integer, Integer> {

    private final int slices;
    private final int samples;

    public PiJob(int slices) {
        this.slices = slices;
        this.samples = (int) Math.min(100000L * slices, Integer.MAX_VALUE);
    }

    @Override
    public Double call(JobContext ctx) throws Exception {
        List<Integer> sampleList = new ArrayList<>();
        for (int i = 0; i < samples; i++) {
            sampleList.add(i);
        }

        return 4.0d * ctx.sc().parallelize(sampleList, slices).map(this).reduce(this) / samples;
    }

    @Override
    public Integer call(Integer v1) {
        double x = Math.random() * 2 - 1;
        double y = Math.random() * 2 - 1;
        return (x * x + y * y < 1) ? 1 : 0;
    }

    @Override
    public Integer call(Integer v1, Integer v2) {
        return v1 + v2;
    }
}